lls/main.c

```diff
@@ -35,15 +48,53 @@ get_lls_conf(void)
 static int
 cleanup_lls(void)
 {
-	/* TODO Destroy ARP and ND cache. */
+	lls_cache_destroy(&lls_conf.arp_cache);
+	/* TODO Destroy ND cache. */
 	rte_timer_stop(&lls_conf.timer);
 	return 0;
 }
 
+int
+arp_resolve(struct in_addr *ipv4_addr, struct ether_addr *eth_addr)
+{
+	return lls_cache_get(&lls_conf.arp_cache, ipv4_addr, eth_addr);
+}
+
+int
+arp_hold(struct in_addr *ipv4_addr, uint8_t port_id)
+{
+	return lls_cache_refcnt(&lls_conf.arp_cache, ipv4_addr, port_id, true);
+}
+
+int
+arp_put(struct in_addr *ipv4_addr, uint8_t port_id)
+{
+	return lls_cache_refcnt(&lls_conf.arp_cache, ipv4_addr, port_id, false);
+}
+
+static inline size_t
+get_vlan_offset(struct ether_hdr *eth_hdr, uint16_t *proto)
+{
+	size_t vlan_offset = 0;
+	if (rte_cpu_to_be_16(ETHER_TYPE_VLAN) == *proto) {
+		struct vlan_hdr *vlan_hdr = (struct vlan_hdr *)(eth_hdr + 1);
+
+		vlan_offset = sizeof(struct vlan_hdr);
+		*proto = vlan_hdr->eth_proto;
+
+		if (rte_cpu_to_be_16(ETHER_TYPE_VLAN) == *proto) {
```

Comment:
Gatekeeper assumes that there's at most one VLAN header since it only knows about at most one VLAN ID. So we can drop packets with more than a VLAN header. Issuing a warning is advisable.

If we ever receive a request to support multiple VLAN IDs in Gatekeeper, we'll need to change that test and redesign the configuration of Gatekeeper to couple with the multiple VLAN IDs.